File: Data Structures & Algorithms/remove-duplicates-from-sorted-array/submission-0.ts
class Solution {
    /**
     * @param {number[]} nums
     * @return {number}
     */
    removeDuplicates(nums: number[]): number {
        let k = 1;

        let slow = 0;
        let fast = 0;

        while (fast <= nums.length - 1) {
            if (nums[slow] == nums[fast]) {
                fast++;
                continue;
            } else {
                nums[slow + 1] = nums[fast];
                slow++;
                k++;
            }
        }

        return k;
    }
}

File: Data Structures & Algorithms/reorder-linked-list/submission-0.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        slow = head
        fast = head.next

        # After this loop, slow now points to the middle element and fast
        # points to the last element.
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        # second half of the list
        second = slow.next

        # Reverse the second half of the list
        slow.next = None
        prev = None
        while second:
            temp = second.next
            second.next = prev
            prev = second
            second = temp
        
        # Second part has been reverse
        # Now merge the two pars of the list
        first = head # left half
        second = prev # right half

        while second:
            temp1, temp2 = first.next, second.next
            first.next = second
            second.next = temp1
            first = temp1
            second = temp2
